### ISSUE TITLE
SegFault on ECDH key de-serialization.

### ISSUE THREAD

**Issue by melvyniandrag** [openssl/openssl#4279]
I cannot find a working reference explaining how to exchange public ecdh keys. The following code is from two days of browsing forums, documentation, and SO. I'm getting a segfault when calling the EC_POINT_oct2point(). Could anyone offer some insight? 

I use shared memory with mmap to share the public keys - I couldn't find any explanation on how to do this except a hint provided [here](https://stackoverflow.com/questions/18155559/how-does-one-access-the-raw-ecdh-public-key-private-key-and-params-inside-opens). 

I made the keys using the code found in the docs. 

```

#include <cassert>
#include <cstring>
#include <iostream>
#include <openssl/evp.h>
#include <openssl/ec.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <unistd.h>

void* create_shared_memory(size_t size) {
  int protection = PROT_READ | PROT_WRITE;
  int visibility = MAP_ANONYMOUS | MAP_SHARED;
  return mmap(NULL, size, protection, visibility, 0, 0);
}

void handleErrors(){
    std::cout << "Null pointer found!" << std::endl;
    exit(1);
}

void ecdh(size_t *secret_len)
{
    const size_t KeySize = 65;
    std::cout << "Allocating " << KeySize << " bytes of shared memory." << std::endl;
    void* shmem = create_shared_memory( KeySize );
    
    int pid = fork();
    EVP_PKEY_CTX *pctx;
    EVP_PKEY_CTX *kctx;
    EVP_PKEY_CTX *ctx;
    unsigned char *secret = NULL;
    EVP_PKEY *pkey = NULL;
    EVP_PKEY *peerkey;
    EVP_PKEY *params = NULL;

    if(NULL == (pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL)))
        handleErrors();

    if(1 != EVP_PKEY_paramgen_init(pctx)) handleErrors();

    if(1 != EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, NID_X9_62_prime256v1)) handleErrors();

    if (!EVP_PKEY_paramgen(pctx, &params)) handleErrors();

    if(NULL == (kctx = EVP_PKEY_CTX_new(params, NULL))) handleErrors();

    if(1 != EVP_PKEY_keygen_init(kctx)) handleErrors();
    if (1 != EVP_PKEY_keygen(kctx, &pkey)) handleErrors();

    const EC_KEY *step1 = EVP_PKEY_get1_EC_KEY( pkey );
    const EC_POINT *step2 = EC_KEY_get0_public_key( step1 ); 
    const EC_GROUP *gp = EC_KEY_get0_group(step1);
    const int UselessPlaceholder = 0;
    unsigned char * SerializedKey = NULL;
    unsigned char * RecvSerializedKey = NULL;
    size_t sizeOfBuffer = EC_POINT_point2oct(gp, step2, POINT_CONVERSION_UNCOMPRESSED, SerializedKey, UselessPlaceholder, NULL); 
    std::cout << "Size of buffer : " << sizeOfBuffer << std::endl;
    if( NULL == ( SerializedKey = static_cast<unsigned char*>( OPENSSL_malloc(sizeOfBuffer)) )) handleErrors();
    if( NULL == ( RecvSerializedKey = static_cast<unsigned char*>( OPENSSL_malloc(sizeOfBuffer)) )) handleErrors();
    const size_t ConfirmBufferSize = EC_POINT_point2oct(gp, step2, POINT_CONVERSION_UNCOMPRESSED, SerializedKey, sizeOfBuffer, NULL); 
    assert(sizeOfBuffer == 65 );
    assert( ConfirmBufferSize == sizeOfBuffer );
    if (pid == 0)
    {
        memcpy( shmem, SerializedKey, sizeOfBuffer );
        std::cout << "Parent Sent: " << SerializedKey << "\n";
        sleep(4);
    }
    else
    {
        sleep(4);
        memcpy( RecvSerializedKey, shmem, sizeOfBuffer );
        std::cout << "Child Received: " << RecvSerializedKey << "\n";
    }
   
    sleep(5);
    
    if (pid != 0)
    {
        memcpy( shmem, SerializedKey, sizeOfBuffer );
        std::cout << "Child Sent: " << SerializedKey << "\n";
        sleep(2);
    }
    else
    {
        sleep(2);
        memcpy( RecvSerializedKey, shmem, sizeOfBuffer );
        std::cout << "Parent Received: " << RecvSerializedKey << "\n";
    }
    
    EC_POINT *deserialPt = NULL;
    int ret1 = EC_POINT_oct2point( gp, deserialPt, RecvSerializedKey, sizeOfBuffer, NULL );
    EC_KEY *deserialKey;
    int ret2 = EC_KEY_set_public_key( deserialKey, deserialPt );
    int ret3 = EVP_PKEY_set1_EC_KEY( peerkey, deserialKey );
    
    if(NULL == (ctx = EVP_PKEY_CTX_new(pkey, NULL))) handleErrors();
    if(1 != EVP_PKEY_derive_init(ctx)) handleErrors();
    if(1 != EVP_PKEY_derive_set_peer(ctx, peerkey)) handleErrors();
    if(1 != EVP_PKEY_derive(ctx, NULL, secret_len)) handleErrors();
    if(NULL == (secret = static_cast<unsigned char*>( OPENSSL_malloc(*secret_len)) )) handleErrors();
    if(1 != (EVP_PKEY_derive(ctx, secret, secret_len))) handleErrors();

    EVP_PKEY_CTX_free(ctx);
    EVP_PKEY_free(peerkey);
    EVP_PKEY_free(pkey);
    EVP_PKEY_CTX_free(kctx);
    EVP_PKEY_free(params);
    EVP_PKEY_CTX_free(pctx);
    
    std::cout << pid << " made secret : " << secret << std::endl;
}

int main(int argc, char** argv){
    size_t secretLen = 256;
    ecdh( &secretLen );
    std::cout << secretLen << std::endl;
}

```

**Comment by melvyniandrag** [comment]
Actually I almost have the damned thing working. This is a nightmare! I wish I was smart enough / with enough free time to write this algorithm myself. At the end of this I have no guarantee that the key transmission even works. I have a "1" return code every time I run it, but honestly it's just been two  days of swearing and looking through header files just to see what commands are available and trying all of them until something works. Update soon. And I guess this is the wrong forum, I head over to SO and reddit to discuss this code sample.

**Comment by snhenson** [comment]
You might be better trying openssl-users. FYI if you just want to serialise the key and don't care on the format it is *much* simpler. For example you can use i2d_PUBKEY/d2i_PUBKEY to serialise the public key of any algorithm.

**Comment by melvyniandrag** [comment]
Ill look into these functions, thanks for the tip.

 For now I have a working example. I have alot of questions about why certain things worked, but other things segfaulted. Somethings could be initialized to NULL, but other things had to be initialized via some other means. There is a cryptic message at the top of the [tutorial](https://wiki.openssl.org/index.php/Elliptic_Curve_Diffie_Hellman) that says "/* NB: assumes pkey, peerkey have been already set up */". As I worked on making my example I found that NULL was ok for some objects, but not for others. And it wasn't clear what "already set up" meant. 

I'm linking the example I wrote in case anyone ever comes across this post at least they'll have a starting point. https://gist.github.com/melvyniandrag/c8bfb6ea6f10f648381ccedcc1ac0912
